File: LeetCode/Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if (len(p) > len(s)):
            return []
        
        sm, pm = {}, {}
        ans = []
        
        for letter in p:
            pm[letter] = pm.get(letter, 0) + 1
        
        l = 0
        
        for r in range(len(s)):
            sm[s[r]] = sm.get(s[r], 0) + 1
            
            while r-l+1 > len(p):
                sm[s[l]] = sm.get(s[l], 1) - 1
                
                if (sm[s[l]] == 0):
                    sm.pop(s[l], -1)
                    
                l += 1
            
            if (sm == pm):
                ans.append(l)
        
        return ans
            

File: LeetCode/Find All Anagrams in a String.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if (p.size() > s.size()) return {};
        
        vector<int> sm(26, 0), pm(26, 0);
        vector<int> ans;
        
        for(int i=0; i < p.size(); i++) {
            pm[p[i] - 'a']++;
        }
        
        int l=0;
        
        for(int r=0; r < s.size(); r++) {
            sm[s[r] - 'a']++;
            
            while(r-l+1 > p.size()) {
                sm[s[l++] - 'a']--;
            }
            
            bool ok = true;
            
            for(int i=0; i < 26; i++) {
                ok &= (sm[i] == pm[i]);
            }
            
            if (ok) {
                ans.push_back(l);
            }
        }
        
        return ans;
    }
};